<launch>
  <arg name="map_yaml"                doc="filepath to previously captured map (required)" />
  <arg name="rviz"     default="true" doc="whether to show Rviz" />
  <arg name="skill_name"  default="pick_pantry" />
  <arg name="model_type"  default="visuomotor_bc" />
  <arg name="train_type"  default="end-eff-img" />
  <arg name="debug_directory" value="$(env HELLO_FLEET_PATH)/debug/"/>
  <arg name="calibrated_controller_yaml_file" value="$(find stretch_core)/config/controller_calibration_head.yaml"/>
  <!-- STRETCH DRIVER -->
  <param name="robot_description" textfile="$(find stretch_description)/urdf/stretch.urdf" />
  <param name="/stretch_driver/broadcast_odom_tf" type="bool" value="true"/> 
  <param name="/stretch_driver/mode" type="string" value="navigation" />
  <param name="/stretch_driver/fail_out_of_range_goal" type="bool" value="false"/>
  <include file="$(find stretch_core)/launch/stretch_driver.launch" pass_all_args="true"/>
  <!-- REALSENSE D435i -->
  <!-- <include file="$(find stretch_core)/launch/d435i_high_resolution.launch"></include>
  <node name="d435i_configure" pkg="stretch_core" type="d435i_configure" output="screen"> -->
    <!--<param name="initial_mode" type="string" value="Default"/>-->
    <!-- <param name="initial_mode" type="string" value="High Accuracy"/>
  </node> -->
  <!-- -->
  <!-- CENTERED BASE LINK -->
  <node name="centered_base_link_tf_publisher" pkg="tf" type="static_transform_publisher" args="-0.1 0 0 0 0 0 1 /base_link /centered_base_link 100" />
  <!-- LASER RANGE FINDER -->
  <include file="$(find stretch_core)/launch/rplidar.launch" />
  <!-- LASER SCAN MATCHER FOR ODOMETRY -->
  <include file="$(find stretch_core)/launch/stretch_scan_matcher.launch" />
  <!-- -->
  <!-- MAP SERVER -->
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_yaml)" />

  <node name="imu_filter_madgwick" pkg="imu_filter_madgwick" type="imu_filter_node" output="screen" respawn="false">
    <param name="use_mag" value="true" />
    <remap from="imu/data_raw" to="imu_mobile_base" />
    <remap from="imu/mag" to="magnetometer_mobile_base" />
  </node>
  <!-- LOCALIZATION
  <param name="/amcl/initial_pose_x" type="double" value="0"/>
  <param name="/amcl/initial_pose_y" type="double" value="0"/>
  <param name="/amcl/initial_pose_a" type="double" value="0"/>
  <include file="$(find amcl)/examples/amcl_diff.launch" /> -->
  <include file="$(find stretch_core)/launch/stretch_ekf.launch"></include>
  <!-- LOCALIZATION -->
  <node pkg="amcl" type="amcl" name="amcl" output="screen">
    <!-- <param name="/amcl/initial_pose_x" type="double" value="1.3438"/>
    <param name="/amcl/initial_pose_y" type="double" value="-0.1"/>
    <param name="/amcl/initial_pose_a" type="double" value="1.57"/>
     -->

    <param name="/amcl/initial_pose_x" type="double" value="0"/>
    <param name="/amcl/initial_pose_y" type="double" value="0"/>
    <param name="/amcl/initial_pose_a" type="double" value="0"/>
    
    <param name="odom_model_type" value="diff-corrected"/>
    <param name="gui_publish_rate" value="10.0"/>
    <param name="laser_max_beams" value="60"/>
    <param name="min_particles" value="500"/>
    <param name="max_particles" value="2000"/>
    <param name="kld_err" value="0.05"/>
    <param name="kld_z" value="0.99"/>
    <param name="odom_alpha1" value="0.045"/>
    <param name="odom_alpha2" value="0.045"/>
    <param name="robot_model_type" value="differential"/>
    <param name="odom_alpha3" value="0.045"/>
    <param name="odom_alpha4" value="0.045"/>
    <param name="laser_z_hit" value="0.5"/>
    <param name="laser_z_short" value="0.05"/>
    <param name="laser_z_max" value="0.05"/>
    <param name="laser_z_rand" value="0.5"/>
    <param name="laser_sigma_hit" value="0.2"/>
    <param name="laser_lambda_short" value="0.1"/>
    <param name="laser_model_type" value="likelihood_field_prob"/>
    <param name="laser_likelihood_max_dist" value="2.0"/>
    <param name="update_min_d" value="0.2"/>
    <param name="update_min_a" value="0.25"/>
    <param name="odom_frame_id" value="odom"/>
    <param name="resample_interval" value="1"/>
    <param name="selective_resampling" value="false"/>
    <param name="transform_tolerance" value="0.1"/>
    <param name="recovery_alpha_slow" value="0.0001"/>
    <param name="recovery_alpha_fast" value="0.01"/>

    <remap from="scan" to="scan_filtered" />
  </node>
  <!-- NAVIGATION -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <rosparam file="$(find stretch_navigation)/config/common_costmap_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find stretch_navigation)/config/common_costmap_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find stretch_navigation)/config/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find stretch_navigation)/config/global_costmap_params_withmap.yaml" command="load" />
    <rosparam file="$(find stretch_navigation)/config/base_local_planner_params.yaml" command="load" />
    <remap from="/cmd_vel" to="/stretch/cmd_vel" />
  </node>
  <!-- -->
  <!-- VISUALIZE -->
  <node name="rviz" pkg="rviz" type="rviz" output="log" args="-d $(find stretch_navigation)/rviz/navigation.rviz" if="$(arg rviz)" />
  <!-- HAL SKLIL -->
</launch>